src/main/java/com/google/devtools/build/lib/analysis/RepoMappingManifestAction.java

```diff
@@ -0,0 +1,147 @@
+import com.google.devtools.build.lib.cmdline.RepositoryName;
+import com.google.devtools.build.lib.collect.nestedset.NestedSet;
+import com.google.devtools.build.lib.collect.nestedset.NestedSetBuilder;
+import com.google.devtools.build.lib.collect.nestedset.Order;
+import com.google.devtools.build.lib.packages.Package;
+import com.google.devtools.build.lib.util.Fingerprint;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import net.starlark.java.eval.EvalException;
+
+/** Creates a manifest file describing the repos and mappings relevant for a runfile tree. */
+public class RepoMappingManifestAction extends AbstractFileWriteAction {
+  private static final UUID MY_UUID = UUID.fromString("458e351c-4d30-433d-b927-da6cddd4737f");
+
+  private final NestedSet<Package> transitivePackages;
+  private final NestedSet<Artifact> runfilesArtifacts;
+  private final String workspaceName;
+
+  public RepoMappingManifestAction(
+      ActionOwner owner,
+      Artifact output,
+      NestedSet<Package> transitivePackages,
+      NestedSet<Artifact> runfilesArtifacts,
+      String workspaceName) {
+    super(owner, NestedSetBuilder.emptySet(Order.STABLE_ORDER), output, /*makeExecutable=*/ false);
+    this.transitivePackages = transitivePackages;
+    this.runfilesArtifacts = runfilesArtifacts;
+    this.workspaceName = workspaceName;
+  }
+
+  @Override
+  public String getMnemonic() {
+    return "RepoMappingManifest";
+  }
+
+  @Override
+  protected String getRawProgressMessage() {
+    return "writing repo mapping manifest for " + getOwner().getLabel();
+  }
+
+  @Override
+  protected void computeKey(
+      ActionKeyContext actionKeyContext,
+      @Nullable ArtifactExpander artifactExpander,
+      Fingerprint fp)
+      throws CommandLineExpansionException, EvalException, InterruptedException {
+    fp.addUUID(MY_UUID);
+    actionKeyContext.addNestedSetToFingerprint(fp, transitivePackages);
+    actionKeyContext.addNestedSetToFingerprint(fp, runfilesArtifacts);
+    fp.addString(workspaceName);
+  }
+
+  @Override
+  public DeterministicWriter newDeterministicWriter(ActionExecutionContext ctx)
+      throws InterruptedException, ExecException {
+    return out -> {
+      Writer writer = new BufferedWriter(new OutputStreamWriter(out, ISO_8859_1));
+
+      ImmutableSet<RepositoryName> reposContributingRunfiles =
+          runfilesArtifacts.toList().stream()
+              .filter(a -> a.getOwner() != null)
+              .map(a -> a.getOwner().getRepository())
+              .collect(toImmutableSet());
+      Map<RepositoryName, RepositoryMapping> reposAndMappings = new HashMap<>();
+      for (Package pkg : transitivePackages.toList()) {
+        reposAndMappings.putIfAbsent(
```

Comment:
Could you add a comment explaining why this is `putIfAbsent` and not `put`? I would assume that there is no difference, is that correct?

Reply:
there's no difference. I can change it to `put` if you prefer that, I don't think it matters either way.

Reply:
Maybe just add a comment saying that repo mappings are uniquely determined by the package and thus this doesn't matter.